src/providers/wfs/qgsoapifprovider.cpp

```diff
@@ -0,0 +1,789 @@
+    if ( mSubsetString.isEmpty() )
+    {
+      mShared->setFeatureCount( itemsRequest.numberMatched(), true );
+    }
+  }
+
+  if ( mShared->mCapabilityExtent.isNull() )
+  {
+    mShared->mCapabilityExtent = itemsRequest.bbox();
+  }
+
+  mShared->mFields = itemsRequest.fields();
+  mShared->mWKBType = itemsRequest.wkbType();
+
+  return true;
+}
+
+void QgsOapifProvider::pushErrorSlot( const QString &errorMsg )
+{
+  pushError( errorMsg );
+}
+
+QgsAbstractFeatureSource *QgsOapifProvider::featureSource() const
+{
+  return new QgsBackgroundCachedFeatureSource( mShared );
+}
+
+QgsFeatureIterator QgsOapifProvider::getFeatures( const QgsFeatureRequest &request ) const
+{
+  return QgsFeatureIterator( new QgsBackgroundCachedFeatureIterator( new QgsBackgroundCachedFeatureSource( mShared ), true, mShared, request ) );
+}
+
+QgsWkbTypes::Type QgsOapifProvider::wkbType() const
+{
+  return mShared->mWKBType;
+}
+
+long QgsOapifProvider::featureCount() const
+{
+  if ( mUpdateFeatureCountAtNextFeatureCountRequest )
+  {
+    mUpdateFeatureCountAtNextFeatureCountRequest = false;
+
+    QgsFeature f;
+    QgsFeatureRequest request;
+    request.setNoAttributes();
+    auto iter = getFeatures( request );
+    int count = 0;
+    bool countExact = true;
+    while ( iter.nextFeature( f ) )
+    {
+      if ( count == 1000 ) // to avoid too long processing time
+      {
+        countExact = false;
+        break;
+      }
+      count ++;
+    }
+
+    mShared->setFeatureCount( count, countExact );
+  }
+  return mShared->getFeatureCount();
+}
+
+QgsFields QgsOapifProvider::fields() const
+{
+  return mShared->mFields;
+}
+
+QgsCoordinateReferenceSystem QgsOapifProvider::crs() const
+{
+  return mShared->mSourceCrs;
+}
+
+QgsRectangle QgsOapifProvider::extent() const
+{
+  return mShared->consolidatedExtent();
+}
+
+void QgsOapifProvider::reloadData()
+{
+  mShared->invalidateCache();
+  QgsVectorDataProvider::reloadData();
+}
+
+bool QgsOapifProvider::isValid() const
+{
+  return mValid;
+}
+
+QgsVectorDataProvider::Capabilities QgsOapifProvider::capabilities() const
+{
+  return QgsVectorDataProvider::SelectAtId | QgsVectorDataProvider::ReadLayerMetadata;
+}
+
+bool QgsOapifProvider::empty() const
+{
+  if ( subsetString().isEmpty() && mShared->isFeatureCountExact() )
+  {
+    return mShared->getFeatureCount( false ) == 0;
+  }
+
+  QgsFeature f;
+  QgsFeatureRequest request;
+  request.setNoAttributes();
+  request.setFlags( QgsFeatureRequest::NoGeometry );
+
+  // Whoops, the provider returns an empty iterator when we are using
+  // a setLimit call in combination with a subsetString.
+  // Remove this method (and default to the QgsVectorDataProvider one)
+  // once this is fixed
+#if 0
+  request.setLimit( 1 );
+#endif
+  return !getFeatures( request ).nextFeature( f );
+
+};
+
+bool QgsOapifProvider::setSubsetString( const QString &sql, bool updateFeatureCount )
+{
+  QgsDebugMsgLevel( QStringLiteral( "sql = '%1'" ).arg( sql ), 4 );
```

Comment:
this isn't really SQL, is it?

Reply:
right. renaming variable and debug message to filter